cmd/synchronization-controller/synchronization-controller.go

```diff
@@ -0,0 +1,362 @@
+	clientTLSConfig       *tls.Config
+	consoleServerPort     int
+	clientcertmanager     certificate.Manager
+	servercertmanager     certificate.Manager
+	clusterConfig         *virtconfig.ClusterConfig
+	reloadableRateLimiter *ratelimiter.ReloadableRateLimiter
+	caManager             kvtls.ClientCAManager
+
+	ctx context.Context
+}
+
+func (app *synchronizationControllerApp) prepareCertManager() (err error) {
+	app.clientcertmanager = bootstrap.NewFileCertificateManager(app.clientCertFilePath, app.clientKeyFilePath)
+	app.servercertmanager = bootstrap.NewFileCertificateManager(app.serverCertFilePath, app.serverKeyFilePath)
+	return
+}
+
+// Update synchronization controller log verbosity on relevant config changes
+func (app *synchronizationControllerApp) shouldChangeLogVerbosity() {
+	verbosity := app.clusterConfig.GetVirtSynchronizationControllerVerbosity()
+	err := log.Log.SetVerbosityLevel(int(verbosity))
+	if err != nil {
+		log.Log.Errorf("unable to change log verbosity to %d, %v", verbosity, err)
+	} else {
+		log.Log.V(2).Infof("set verbosity to %d", verbosity)
+	}
+}
+
+// Update synchronization controller rate limiter
+func (app *synchronizationControllerApp) shouldChangeRateLimiter() {
+	config := app.clusterConfig.GetConfig()
+	qps := config.HandlerConfiguration.RestClient.RateLimiter.TokenBucketRateLimiter.QPS
+	burst := config.HandlerConfiguration.RestClient.RateLimiter.TokenBucketRateLimiter.Burst
+	app.reloadableRateLimiter.Set(flowcontrol.NewTokenBucketRateLimiter(qps, burst))
+	log.Log.V(2).Infof("setting rate limiter to %v QPS and %v Burst", qps, burst)
+}
+
+func (app *synchronizationControllerApp) setupTLS(factory controller.KubeInformerFactory) error {
+	kubevirtCAConfigInformer := factory.KubeVirtCAConfigMap()
+	if err := kubevirtCAConfigInformer.SetWatchErrorHandler(func(r *cache.Reflector, err error) {
+		apiHealthVersion.Clear()
+		cache.DefaultWatchErrorHandler(r, err)
+	}); err != nil {
+		return err
+	}
+	app.caManager = kvtls.NewCAManager(kubevirtCAConfigInformer.GetStore(), app.namespace, app.caConfigMapName)
+
+	app.serverTLSConfig = kvtls.SetupTLSForVirtSynchronizationControllerServer(app.caManager, app.servercertmanager, app.externallyManaged, app.clusterConfig)
+	app.clientTLSConfig = kvtls.SetupTLSForVirtSynchronizationControllerClients(app.caManager, app.clientcertmanager, app.externallyManaged)
+
+	return nil
+}
+
+func (app *synchronizationControllerApp) Run() {
+	var err error
+
+	app.LeaderElection = leaderelectionconfig.DefaultLeaderElectionConfiguration()
+
+	app.reloadableRateLimiter = ratelimiter.NewReloadableRateLimiter(flowcontrol.NewTokenBucketRateLimiter(virtconfig.DefaultVirtControllerQPS, virtconfig.DefaultVirtHandlerBurst))
+	clientConfig, err := kubecli.GetKubevirtClientConfig()
+	if err != nil {
+		panic(err)
+	}
+	clientConfig.RateLimiter = app.reloadableRateLimiter
+	app.virtCli, err = kubecli.GetKubevirtClientFromRESTConfig(clientConfig)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
maybe we should do a backoff here?
I'm just thinking, what if the API server is not immediately available, we will crashloop, but I guess we can do better - it's not a must but would add to resiliancy